### PR TITLE
[terra-dropdown]Added aria-haspoup and aria-controls attribute

### PR DESCRIPTION
**Pull request by AH106586Harika**
### Summary
fixed- Screenreader not communicating menu

**What was changed:**
Added aria-haspoup and aria-controls attribute.

**Why it was changed:**
The screen reader is not announcing that it has a menu when the drop down button is navigation

### Testing

This change was tested using:

- [ ] WDIO
- [X] Jest
- [ ] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:

- [ ] UX review
- [ ] Accessibility review
- [ ] Functional review

### Additional Details

**This PR resolves:**

UXPLATFORM-9694

---

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

**Comment by sdadn** [comment]
Is this PR ready for review?

**Comment by AH106586Harika** [comment]
> Is this PR ready for review?

Yes.

**Comment by supreethmr** [comment]
@AH106586Harika  on which Screen reader this issue was identified ?

File: packages/terra-dropdown-button/CHANGELOG.md
Comment (supreethmr) [inline packages/terra-dropdown-button/CHANGELOG.md:5]:
```suggestion

* Added
  * Added `aria-haspoup` and `aria-controls` attributes for dropdown button.
```

File: packages/terra-dropdown-button/src/_DropdownList.jsx
Comment (supreethmr) [inline packages/terra-dropdown-button/src/_DropdownList.jsx:230]: why do we need aria-LabelledBy for `ul` menu ?

File: packages/terra-dropdown-button/src/_Dropdown.jsx
Comment (supreethmr) [inline packages/terra-dropdown-button/src/_Dropdown.jsx:80]: can we store them in const and use it here ??
`const { menuId, buttonId } = customProps;`

File: packages/terra-dropdown-button/tests/jest/DropdownButton.test.jsx
Comment (supreethmr) [inline packages/terra-dropdown-button/tests/jest/DropdownButton.test.jsx:101]: can we have jest test to ensure button is assigned with value passed by consumer as ID instead of the auto generated ID.

**Comment by supreethmr** [comment]
Since UUID is now part of component, We need to update the About page for dropdown button and split button to include testing details similar to this : https://github.com/cerner/terra-framework/blob/4ce7712fde9ecf6235caceb0115c6de62dbb4a68/packages/terra-framework-docs/src/terra-dev-site/doc/menu/Menu.1.doc.mdx#L67C1-L67C1 .

and we need to update react-intl as peer dependency along with implementation instructions like this : https://github.com/cerner/terra-core/pull/3854/files#diff-43d71167fbce146e41d65538d63d237642369f5d98c53c55d8294d764b82ddfb

File: packages/terra-dropdown-button/src/_DropdownList.jsx
Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:230]: As suggested by UX team and provided the reference URL to add aria-labelledby for dropdown menu

https://www.w3.org/WAI/ARIA/apg/patterns/menu-button/examples/menu-button-links/

**Comment by AH106586Harika** [comment]
> @AH106586Harika on which Screen reader this issue was identified ?

As requested by UX team, we have made this changes as per standards, Please find the more details on this jira [UXPLATFORM-9694](https://jira2.cerner.com/browse/UXPLATFORM-9694)

**Comment by AH106586Harika** [comment]
> Since UUID is now part of component, We need to update the About page for dropdown button and split button to include testing details similar to this : https://github.com/cerner/terra-framework/blob/4ce7712fde9ecf6235caceb0115c6de62dbb4a68/packages/terra-framework-docs/src/terra-dev-site/doc/menu/Menu.1.doc.mdx#L67C1-L67C1 .
> 
> and we need to update react-intl as peer dependency along with implementation instructions like this : https://github.com/cerner/terra-core/pull/3854/files#diff-43d71167fbce146e41d65538d63d237642369f5d98c53c55d8294d764b82ddfb


Done.

File: packages/terra-dropdown-button/tests/jest/DropdownButton.test.jsx
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/jest/DropdownButton.test.jsx:101]: Done.

File: packages/terra-dropdown-button/CHANGELOG.md
Comment (AH106586Harika) [reply packages/terra-dropdown-button/CHANGELOG.md:5]: Done.

File: packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/About.1.doc.mdx
Comment (supreethmr) [inline packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/About.1.doc.mdx:73]:
```suggestion
Terra Dropdown button uses `uuid` which changes the component's description id dynamically. To mock the return value with the Jest testing library, `jest.spyOn` can be used.
```

File: packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/SplitButton.2.doc.mdx
Comment (supreethmr) [inline packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/SplitButton.2.doc.mdx:58]:
```suggestion
Terra Split Button uses `uuid` which changes the component's description id dynamically. To mock the return value with the Jest testing library, `jest.spyOn` can be used.
```

File: packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/About.1.doc.mdx
Comment (AH106586Harika) [reply packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/About.1.doc.mdx]: Done.

File: packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/SplitButton.2.doc.mdx
Comment (AH106586Harika) [reply packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/SplitButton.2.doc.mdx]: Done.

File: packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/About.1.doc.mdx
Comment (supreethmr) [inline packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/About.1.doc.mdx:43]: this should point to application-base and we need to update react-intl as well on peer dependency table

Comment (AH106586Harika) [reply packages/terra-core-docs/src/terra-dev-site/doc/dropdown-button/About.1.doc.mdx:43]: Updated URL to application base and already react-intl is there in peer dependency table.

image

**Review by supreethmr — APPROVED** [review]
Please get approval from A11y Team before PR merge

File: packages/terra-dropdown-button/src/_DropdownList.jsx
Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:230]: I had a look on the button navigation and it is not announcing text of button on menu change, it is reading as 'menu label (1 of 4) menu item' please validate at your end as well if required.

Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:230]: https://github.com/cerner/terra-core/assets/126655256/f73a9f29-1779-4b3f-bcd1-4405fc11da81

Comment (supreethmr) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:230]: you can notice the reported behavior on the screen recording that VO did announced button text as `menu Export, Deafult dropdown` which is not correct for the first dropdown item

Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/_DropdownList.jsx]: Updated as discussed.

File: packages/terra-dropdown-button/src/DropdownButton.jsx
Comment (supreethmr) [inline packages/terra-dropdown-button/src/DropdownButton.jsx:221]: we need to add aria-controls for split button as well : https://github.com/cerner/terra-core/blob/7cb2fd38b0daa2deb8fb1a72d49d9502c0fdbebe/packages/terra-dropdown-button/src/SplitButton.jsx#L281

Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/DropdownButton.jsx:221]: Implemented.

File: packages/terra-dropdown-button/tests/jest/DropdownButton.test.jsx
Comment (supreethmr) [inline packages/terra-dropdown-button/tests/jest/DropdownButton.test.jsx:18]: Need to add similar spy for splitbutton test as well

Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/jest/DropdownButton.test.jsx:18]: Done.

**Comment by rahulkumar8cs** [comment]
+1

**Review by sugan2416 — APPROVED** [review]

**Pull request merged**
